### PR TITLE
[WIP] Separate optimizer implementations

### PR DESCRIPTION
**Pull request by myui**
This PR is based on [a pending PR](https://github.com/myui/hivemall/pull/285) by @maropu that is sent before Hivemall entered Apache Incubator.

See [JIRA](https://issues.apache.org/jira/browse/HIVEMALL-22
) for tracking the status of this issue.

### CODE REVIEW

File: core/src/main/java/hivemall/model/NewDenseModel.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+        int i = HiveUtils.parseInt(feature);
+        if (i >= size) {
+            return 1f;
+        }
+        return covars[i];
+    }
+
+    @Override
+    protected void _set(Object feature, float weight, short clock) {
+        int i = ((Integer) feature).intValue();
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/model/NewDenseModel.java]: Unboxing isn't necessary?

File: core/src/main/java/hivemall/model/NewDenseModel.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean contains(Object feature) {
+        int i = HiveUtils.parseInt(feature);
+        if (i >= size) {
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/model/NewDenseModel.java]: Should we also check negative value? `feature` must be 0~size.

File: core/src/main/java/hivemall/model/NewDenseModel.java [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+            return false;
+        }
+        float w = weights[i];
+        return w != 0.f;
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/model/NewDenseModel.java]: Using `Float.compare` can be safer.

File: core/src/main/java/hivemall/model/NewSpaceEfficientDenseModel.java [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        deltaUpdates[feature] = 0;
+    }
+
+    private float getWeight(final int i) {
+        final short w = weights[i];
+        return (w == HalfFloat.ZERO) ? HalfFloat.ZERO : HalfFloat.halfFloatToFloat(w);
+    }
+
+    private float getCovar(final int i) {
+        return HalfFloat.halfFloatToFloat(covars[i]);
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/model/NewSpaceEfficientDenseModel.java]: covar can be null. So we should check hasCovar first.

File: core/src/main/java/hivemall/model/NewSpaceEfficientDenseModel.java [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+        }
+
+        @Override
+        public Integer getKey() {
+            return cursor;
+        }
+
+        @Override
+        public IWeightValue getValue() {
+            if (covars == null) {
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/model/NewSpaceEfficientDenseModel.java]: How about checking with `hasCovariance`?

File: core/src/main/java/hivemall/model/NewSparseModel.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                    wrapper = new WeightValueWithCovarClock(value);
+                    break;
+                case ParamsF1:
+                    wrapper = new WeightValueParamsF1Clock(value);
+                    break;
+                case ParamsF2:
+                    wrapper = new WeightValueParamsF2Clock(value);
+                    break;
+                default:
+                    throw new IllegalStateException("Unexpected value type: " + value.getType());
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/model/NewSparseModel.java]: Is there any reason that `ParamF3` is not supported?

File: core/src/main/java/hivemall/optimizer/EtaEstimator.java [hunk 7]
```diff
@@ -157,4 +158,34 @@ public static EtaEstimator get(@Nullable CommandLine cl, float defaultEta0)
         return new InvscalingEtaEstimator(eta0, power_t);
     }
 
+    @Nonnull
+    public static EtaEstimator get(@Nonnull final Map<String, String> options)
+            throws IllegalArgumentException {
+        final String etaName = options.get("eta");
+        if(etaName == null) {
+            return new FixedEtaEstimator(1.f);
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/optimizer/EtaEstimator.java]: I think default initial eta value can be same to the default `eta0` value (0.1f). 1.f looks large.

File: core/src/main/java/hivemall/optimizer/Optimizer.java [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import hivemall.model.WeightValue;
+import hivemall.model.IWeightValue;
+
+public interface Optimizer {
+
+    /**
+     * Update the weights of models thru this interface.
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/optimizer/Optimizer.java]: s/thru/through/

File: core/src/main/java/hivemall/optimizer/Optimizer.java [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+     */
+    float computeUpdatedValue(@Nonnull Object feature, float weight, float gradient);
+
+    // Count up #step to tune learning rate
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/optimizer/Optimizer.java]: Javadoc style might be better like `computeUpdatedValue` docs.

File: core/src/main/java/hivemall/optimizer/Optimizer.java [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+        private final float decay;
+        private final float eps;
+        private final float scale;
+
+        public AdaDelta(Map<String, String> options) {
+            super(options);
+            float decay = 0.95f;
+            float eps = 1e-6f;
+            float scale = 100.0f;
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/optimizer/Optimizer.java]: How can you find the usage of `scale` value? I coundn't find that in original paper.
https://arxiv.org/pdf/1212.5701v1.pdf

File: core/src/main/java/hivemall/optimizer/Regularization.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+
+public abstract class Regularization {
+
+    protected final float lambda;
+
+    public Regularization(final Map<String, String> options) {
+        float lambda = 1e-6f;
```
Comment (Lewuathe) [inline core/src/main/java/hivemall/optimizer/Regularization.java]: I'm not sure this is the good value as default for penalty weight.
It looks a bit smaller. sklearn SGD implementation uses 0.0001 as default.
http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html

**Comment by maropu** [comment]
@Lewuathe Thanks for your review! I'll check later

**Comment by myui** [comment]
I'm considering to review and merge this PR after finishing #13 .

**Comment by myui** [comment]
LossFunction should be selectable, not fixed.

File: core/src/main/java/hivemall/model/NewSpaceEfficientDenseModel.java (hunk 4 above)
Comment (Lewuathe) [inline core/src/main/java/hivemall/model/NewSpaceEfficientDenseModel.java]: I see.

File: core/src/main/java/hivemall/optimizer/Optimizer.java (hunk 10 above)
Comment (Lewuathe) [inline core/src/main/java/hivemall/optimizer/Optimizer.java]: I see. Sounds reasonable.

**Comment by takuti** [comment]
@myui @maropu @Lewuathe JFYI: This PR has been moved to #79, and now I'm finalizing this huge feature.

@myui Could you close this PR?

**Comment by myui** [comment]
Moved to #79

**Pull request closed without merging**